typing/uniqueness_analysis.ml

```diff
@@ -1058,117 +1062,133 @@ let rec pattern_match_tuple pat values =
+    either that the memory access is not pushed down or that no destructive
+    updates can be performed on the memory.
+
+    Each pattern falls into one of three cases:
+    - If we do not read from the underlying memory,
+      we do not have to take an action.
+    - We can allow destructive updates later on by borrowing the
+      memory address. Then we have to protect the read from getting
+      pushed down using a unique barrier.
+    - We can disallow any destructive updates following the read
+      by consuming the memory address as aliased. *)
+and pattern_match_barrier pat paths : UF.t =
   let loc = pat.pat_loc in
   let occ = Occurrence.mk loc in
-  (* To read from the allocation, we need to borrow its memory cell
-     and set the unique_barrier. However, we do not read in every case,
-     since the user might want use a wildcard for already-consumed data. *)
-  let no_borrow_memory_address () =
-    Unique_barrier.enable pat.pat_unique_barrier;
-    ignore (Unique_barrier.resolve pat.pat_unique_barrier)
+  Unique_barrier.enable pat.pat_unique_barrier;
+  let no_memory_access () =
+    ignore (Unique_barrier.resolve pat.pat_unique_barrier);
+    UF.unused
   in
   let borrow_memory_address () =
-    Unique_barrier.enable pat.pat_unique_barrier;
     Paths.mark_implicit_borrow_memory_address occ (Read pat.pat_unique_barrier)
       paths
   in
+  let consume_memory_address reason =
+    ignore (Unique_barrier.resolve pat.pat_unique_barrier);
+    Paths.mark_aliased occ reason paths
+  in
   match pat.pat_desc with
-  | Tpat_or (pat0, pat1, _) ->
-    no_borrow_memory_address ();
-    let ext0, uf0 = pattern_match_single pat0 paths in
-    let ext1, uf1 = pattern_match_single pat1 paths in
-    Ienv.Extension.disjunct ext0 ext1, UF.choose uf0 uf1
-  | Tpat_any ->
-    no_borrow_memory_address ();
-    Ienv.Extension.empty, UF.unused
-  | Tpat_var (id, _, _, _) ->
-    no_borrow_memory_address ();
-    Ienv.Extension.singleton id paths, UF.unused
-  | Tpat_alias (pat', id, _, _, _) ->
-    no_borrow_memory_address ();
-    let ext0 = Ienv.Extension.singleton id paths in
-    let ext1, uf = pattern_match_single pat' paths in
-    Ienv.Extension.conjunct ext0 ext1, uf
+  | Tpat_or _ -> no_memory_access ()
+  | Tpat_any -> no_memory_access ()
+  | Tpat_var _ -> no_memory_access ()
+  | Tpat_alias _ -> no_memory_access ()
   | Tpat_constant _ ->
-    let uf_read = borrow_memory_address () in
-    Ienv.Extension.empty, uf_read
-  | Tpat_construct (lbl, cd, pats, _) ->
-    let uf_read = borrow_memory_address () in
-    let pats_args = List.combine pats cd.cstr_args in
-    let ext, uf_pats =
+    (* This is necessary since we can not guarantee that
+       the reads of constants in the pattern-matching code
+       are never pushed down. *)
+    consume_memory_address Constant
+  | Tpat_construct _ -> borrow_memory_address ()
+  | Tpat_variant _ -> borrow_memory_address ()
+  | Tpat_record _ -> borrow_memory_address ()
```

Comment:
Now I'm confused again. Sorry! This doesn't have to read memory to tell whether the pattern succeeds, because a record pattern always succeeds. So I think I just don't understand the property we're trying to capture here.

Reply:
Yes, that is correct. We could create a special case here for record patterns that do not bind any names and those would not require a memory access. But part of the record pattern is that we can learn the memory addresses of its children by naming them which requires dereferencing the pointer to the record.